### PR TITLE
Delay Universe Substitution in CClosure

### PR DESCRIPTION
**Pull request by SkySkimmer**
Return of https://github.com/coq/coq/pull/8747

Co-authored-by: Janno <Janno@users.noreply.github.com>

Overlay: https://github.com/Mtac2/Mtac2/pull/354

### CODE REVIEW

**Comment by SkySkimmer** [comment]
Bench https://gitlab.com/coq/coq/-/jobs/2200140750

File: kernel/cClosure.ml [hunk 1]
```diff
@@ -317,19 +319,19 @@ and fterm =
   | FConstruct of pconstructor
   | FApp of fconstr * fconstr array
   | FProj of Projection.t * fconstr
-  | FFix of fixpoint * fconstr subs
```
Comment (ppedrot) [inline kernel/cClosure.ml:320]: Despite being more annoying to write, I suspect that expanding `subs` here to save an indirection would reduce the runtime overhead observed in mathcomp.

**Comment by ppedrot** [comment]
I am quite surprised that not only category-theory did not experience a speed-up, but actually slowed down. When I checked what this dev was doing, it was really passing its time in universe substitution induced by conversion...

**Comment by ppedrot** [comment]
Do I read correctly the code that we don't use the instance in the reference cache? That would explain why we don't get any speed up when universes matter. This is the critical part, not the other functions.

**Comment by SkySkimmer** [comment]
>Do I read correctly the code that we don't use the instance in the reference cache?

Let's try this again
https://gitlab.com/coq/coq/-/jobs/2204355860

**Comment by SkySkimmer** [comment]
to_constr_case is wrong, probably the entire to_constr/subst_constr code needs to be checked

**Comment by SkySkimmer** [comment]
@coqbot ci minimize

**Comment by SkySkimmer** [comment]
HoTT finally works! I even ran the produced files through master's coqchk and it didn't complain.

**Comment by SkySkimmer** [comment]
https://gitlab.com/coq/coq/-/jobs/2211144947

**Comment by SkySkimmer** [comment]
Needed to fix the caseinvert stuff too

**Comment by SkySkimmer** [comment]
https://gitlab.com/coq/coq/-/jobs/2211513935

**Comment by SkySkimmer** [comment]
~~~
┌──────────────────────────┬──────────────────────────┬────────────────────────────────────────┬────────────────────────────────────────┬─────────────────────────┬───────────────────┐
│                          │      user time [s]       │               CPU cycles               │            CPU instructions            │  max resident mem [KB]  │    mem faults     │
│                          │                          │                                        │                                        │                         │                   │
│       package_name       │   NEW      OLD    PDIFF  │      NEW             OLD        PDIFF  │      NEW             OLD        PDIFF  │   NEW      OLD    PDIFF │ NEW  OLD   PDIFF  │
├──────────────────────────┼──────────────────────────┼────────────────────────────────────────┼────────────────────────────────────────┼─────────────────────────┼───────────────────┤
│      coq-category-theory │ 1030.49  1703.63  -39.51 │  4507541507202   7463396439770  -39.60 │  7473378186105  15541483049816  -51.91 │ 1322440  1469008  -9.98 │   1    0      nan │
│                 coq-hott │  162.89   168.11   -3.11 │   707079354254    730288043510   -3.18 │  1108485829131   1160195873192   -4.46 │  621336   617400   0.64 │   0    0      nan │
│             coq-compcert │  290.87   290.87    0.00 │  1264869203283   1263952090091    0.07 │  1897867702712   1887393846578    0.55 │ 1106120  1105132   0.09 │   0    8  -100.00 │
│   coq-mathcomp-ssreflect │   25.63    25.61    0.08 │   112002354351    111575848962    0.38 │   140660664489    139376307221    0.92 │  538712   537648   0.20 │   0    0      nan │
│      coq-metacoq-erasure │   64.00    63.95    0.08 │   275083026045    274860067591    0.08 │   418787574412    417606986011    0.28 │ 1084000  1090396  -0.59 │   0    0      nan │
│             coq-coqprime │   43.99    43.92    0.16 │   190819133238    190578559009    0.13 │   283213193033    282523565372    0.24 │  763336   761288   0.27 │   0    0      nan │
│            coq-fourcolor │ 1510.26  1507.71    0.17 │  6602358328491   6591128115235    0.17 │ 12082905946485  12067701963950    0.13 │  789944   791720  -0.22 │   0    0      nan │
│              coq-coqutil │   36.08    36.01    0.19 │   155909569289    154866260929    0.67 │   209882188737    208744725000    0.54 │  607508   604080   0.57 │   4    0      nan │
│     coq-metacoq-template │   79.32    79.12    0.25 │   340531817512    338153603682    0.70 │   511257411669    508996278392    0.44 │  932848   942768  -1.05 │   4    0      nan │
│         coq-math-classes │   93.18    92.86    0.34 │   405549715411    403295679134    0.56 │   572284896200    569468893582    0.49 │  502800   503000  -0.04 │   0    0      nan │
│  coq-metacoq-safechecker │  164.01   163.39    0.38 │   714238717230    712155260676    0.29 │   961174352694    955270492700    0.62 │ 1191860  1190308   0.13 │   4    2   100.00 │
│    coq-engine-bench-lite │  183.33   182.58    0.41 │   770692174793    766990147280    0.48 │  1500687316621   1493270135098    0.50 │ 1304160  1234824   5.62 │   0    0      nan │
│               coq-flocq3 │   79.61    79.28    0.42 │   346826767259    345448117239    0.40 │   458929096850    455990719692    0.64 │  991888   991704   0.02 │   0    0      nan │
│                  coq-vst │ 1099.93  1094.95    0.45 │  4806459474655   4780668886909    0.54 │  7786332732578   7711852965257    0.97 │ 2132640  2111624   1.00 │   3    1   200.00 │
│            coq-fiat-core │   57.00    56.74    0.46 │   236088879294    234955915393    0.48 │   330819829269    330017624638    0.24 │  481808   479780   0.42 │   0    6  -100.00 │
│               coq-geocoq │  722.74   719.11    0.50 │  3156633011015   3141296152095    0.49 │  5019372890855   4984035806788    0.71 │ 1075864  1076548  -0.06 │   3    2    50.00 │
│              coq-bignums │   28.78    28.63    0.52 │   125162125496    124885926825    0.22 │   172868888897    172016616256    0.50 │  471064   469096   0.42 │   0    0      nan │
│           coq-coquelicot │   34.59    34.39    0.58 │   149283284990    148228196389    0.71 │   196840563531    195612145539    0.63 │  755324   755504  -0.02 │ 205    0      nan │
│             coq-bedrock2 │  227.83   226.33    0.66 │   994643752741    988781275476    0.59 │  1794582582036   1785296738099    0.52 │ 2051876  2052056  -0.01 │  17   11    54.55 │
│                coq-color │  217.52   216.04    0.69 │   944768733108    937056289340    0.82 │  1330786377283   1324259110468    0.49 │ 1096300  1111020  -1.32 │   0    7  -100.00 │
│          coq-fiat-crypto │ 2925.10  2903.52    0.74 │ 12762536900669  12669678649687    0.73 │ 23309791243281  23215238182723    0.41 │ 3850184  3843096   0.18 │  76   79    -3.80 │
│                 coq-corn │  786.70   780.61    0.78 │  3435498084998   3410130097309    0.74 │  5300727306048   5270865926914    0.57 │  866056   877272  -1.28 │   0    0      nan │
│                coq-verdi │   49.31    48.92    0.80 │   214467028589    212346847644    1.00 │   321302351042    318256684812    0.96 │  840308   836864   0.41 │   0    0      nan │
│         coq-fiat-parsers │  346.43   343.53    0.84 │  1495232534590   1483469543016    0.79 │  2460220085773   2448047952508    0.50 │ 3042172  2694784  12.89 │   2   28   -92.86 │
│    coq-mathcomp-fingroup │   22.33    22.14    0.86 │    97327011287     96832982798    0.51 │   140711520894    139157633794    1.12 │  483504   485704  -0.45 │   0    0      nan │
│           coq-verdi-raft │  582.87   577.40    0.95 │  2546060911603   2520843451840    1.00 │  3966259870543   3925228611702    1.05 │ 1231996  1231776   0.02 │   0    0      nan │
│             coq-rewriter │  386.13   382.50    0.95 │  1685295733433   1668523504191    1.01 │  2787195391910   2770665390120    0.60 │ 1060880  1060184   0.07 │   7    9   -22.22 │
│ coq-metacoq-translations │   20.94    20.74    0.96 │    90768599352     89514341002    1.40 │   157785510835    155647048859    1.37 │  962400   965992  -0.37 │   0    0      nan │
│                 coq-core │  108.30   107.24    0.99 │   389321031385    386863351841    0.64 │   442636651579    442235517669    0.09 │  280956   279564   0.50 │   0    0      nan │
│            coq-perennial │ 4141.05  4100.49    0.99 │ 18107590901454  17932550359457    0.98 │ 30033580174200  29772131149853    0.88 │ 2618768  2624860  -0.23 │   0    0      nan │
│        coq-metacoq-pcuic │  559.18   553.66    1.00 │  2443042798144   2419477696877    0.97 │  3465123015152   3435950829212    0.85 │ 1661716  1662184  -0.03 │  29    1  2800.00 │
│       coq-mathcomp-field │  102.55   101.49    1.04 │   449362081261    444910171969    1.00 │   744896836281    736719951755    1.11 │  706816   708148  -0.19 │   0    0      nan │
│   coq-mathcomp-character │   69.99    69.26    1.05 │   306555578220    303368274593    1.05 │   468382422687    462887684994    1.19 │  829852   825156   0.57 │   0    0      nan │
│        coq-iris-examples │  429.57   424.83    1.12 │  1875018324684   1855557002110    1.05 │  2859141336190   2837330356079    0.77 │ 1179712  1182736  -0.26 │   0    0      nan │
│     coq-mathcomp-algebra │   60.51    59.84    1.12 │   264958965135    261711543805    1.24 │   363223163065    359524766888    1.03 │  598840   595744   0.52 │   0    0      nan │
│              coq-unimath │ 4501.64  4449.17    1.18 │ 19712779847376  19487353328082    1.16 │ 39344690014180  38985788763963    0.92 │ 3269020  3258448   0.32 │   1    0      nan │
│    coq-mathcomp-solvable │   83.24    82.23    1.23 │   364319354354    360068745058    1.18 │   558359047303    551712845001    1.20 │  665040   665204  -0.02 │   0    0      nan │
│               coq-stdlib │  419.89   414.61    1.27 │  1491088568697   1474612437734    1.12 │  1423446996895   1427002595369   -0.25 │  596740   608108  -1.87 │   0    0      nan │
│   coq-mathcomp-odd-order │  545.16   537.15    1.49 │  2389672438790   2355400822763    1.46 │  4089667715589   4048324834136    1.02 │ 1228664  1212308   1.35 │   0    0      nan │
└──────────────────────────┴──────────────────────────┴────────────────────────────────────────┴────────────────────────────────────────┴─────────────────────────┴───────────────────┘
~~~

**Comment by Alizter** [comment]
<details>
<summary>Slowdowns</summary>

```
┌─────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────┐
│                                                          TOP 25 SLOW DOWNS                                                          │
│                                                                                                                                     │
│   OLD       NEW      DIFF    %DIFF    Ln                   FILE                                                                     │
├─────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────┤
│  12.1370   14.4010  2.2640   18.65%  1518  coq-unimath/UniMath/Bicategories/DisplayedBicats/Examples/DisplayedInserter.v.html       │
│ 141.7360  143.2820  1.5460    1.09%    47  coq-fiat-crypto/src/Curves/Weierstrass/AffineProofs.v.html                               │
│ 105.2180  106.7580  1.5400    1.46%    22  coq-fiat-crypto/src/Rewriter/Passes/ArithWithCasts.v.html                                │
│  17.2530   18.4520  1.1990    6.95%   444  coq-perennial/src/program_proof/memkv/memkv_put_proof.v.html                             │
│  16.0840   17.0850  1.0010    6.22%  1640  coq-unimath/UniMath/Bicategories/DisplayedBicats/FiberBicategory/FiberBicategory1.v.html │
│  14.8830   15.8580  0.9750    6.55%   860  coq-perennial/src/program_proof/buf/buf_proof.v.html                                     │
│ 219.5710  220.4020  0.8310    0.38%   139  coq-fiat-crypto/src/UnsaturatedSolinasHeuristics/Tests.v.html                            │
│  20.1800   20.9350  0.7550    3.74%   594  coq-perennial/src/goose_lang/lib/encoding/encoding.v.html                                │
│  56.2970   57.0430  0.7460    1.33%   857  coq-unimath/UniMath/Bicategories/DisplayedBicats/Examples/DisplayedInserter.v.html       │
│  42.7590   43.4400  0.6810    1.59%   833  coq-fiat-crypto/src/Fancy/Compiler.v.html                                                │
│  17.5820   18.2420  0.6600    3.75%   216  coq-unimath/UniMath/Bicategories/DualityInvolution.v.html                                │
│   8.8160    9.4750  0.6590    7.48%    62  coq-unimath/UniMath/Bicategories/DisplayedBicats/DispBuilders.v.html                     │
│  33.8680   34.4880  0.6200    1.83%   352  coq-fiat-crypto/src/Curves/Montgomery/XZProofs.v.html                                    │
│  67.1260   67.7360  0.6100    0.91%    27  coq-fiat-crypto/src/Rewriter/Passes/ToFancyWithCasts.v.html                              │
│  38.0930   38.7030  0.6100    1.60%   520  coq-perennial/src/program_proof/txn/twophase_refinement_proof.v.html                     │
│  16.6790   17.2810  0.6020    3.61%   283  coq-unimath/UniMath/Bicategories/DualityInvolution.v.html                                │
│  12.4970   13.0690  0.5720    4.58%   900  coq-mathcomp-odd-order/theories/PFsection4.v.html                                        │
│  12.9430   13.4780  0.5350    4.13%  1289  coq-unimath/UniMath/Bicategories/DualityInvolution.v.html                                │
│  18.4850   19.0140  0.5290    2.86%   193  coq-unimath/UniMath/HomologicalAlgebra/KATriangulated.v.html                             │
│   7.6690    8.1600  0.4910    6.40%   858  coq-fiat-crypto/src/Fancy/Compiler.v.html                                                │
│  15.0200   15.5110  0.4910    3.27%   476  coq-unimath/UniMath/HomologicalAlgebra/KA.v.html                                         │
│   1.3990    1.8890  0.4900   35.03%  3266  coq-metacoq-pcuic/pcuic/theories/PCUICConfluence.v.html                                  │
│  12.7950   13.2550  0.4600    3.60%  1108  coq-unimath/UniMath/Bicategories/DisplayedBicats/FiberBicategory/FiberBicategory1.v.html │
│   0.1840    0.6300  0.4460  242.39%   871  coq-perennial/src/program_proof/wal/installer_proof.v.html                               │
│  16.3070   16.7450  0.4380    2.69%   324  coq-unimath/UniMath/Bicategories/DualityInvolution.v.html                                │
└─────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────┘
```

</details>

<details>
<summary>Speedups</summary>

```

┌─────────────────────────────────────────────────────────────────────────────────────────────────────────────────┐
│                                                TOP 25 SPEED UPS                                                 │
│                                                                                                                 │
│   OLD      NEW      DIFF     %DIFF    Ln               FILE                                                     │
├─────────────────────────────────────────────────────────────────────────────────────────────────────────────────┤
│ 50.9770   5.7490  -45.2280  -88.72%    57  coq-category-theory/Construction/Comma/Natural/Transformation.v.html │
│ 20.2250   4.7890  -15.4360  -76.32%   773  coq-category-theory/Construction/Comma/Adjunction.v.html             │
│ 19.4860   4.8010  -14.6850  -75.36%   237  coq-category-theory/Construction/Comma/Adjunction.v.html             │
│ 16.3780   2.4140  -13.9640  -85.26%   320  coq-category-theory/Construction/Comma/Adjunction.v.html             │
│ 18.5870   4.8990  -13.6880  -73.64%   271  coq-category-theory/Construction/Comma/Adjunction.v.html             │
│ 16.0260   3.4620  -12.5640  -78.40%   132  coq-category-theory/Functor/Strong/Product.v.html                    │
│ 16.4080   4.0550  -12.3530  -75.29%   251  coq-category-theory/Construction/Comma/Adjunction.v.html             │
│ 13.9660   3.6560  -10.3100  -73.82%    65  coq-category-theory/Construction/Comma/Natural/Transformation.v.html │
│ 16.8310   7.3470   -9.4840  -56.35%    90  coq-category-theory/Functor/Strong/Product.v.html                    │
│ 12.3760   3.0400   -9.3360  -75.44%   717  coq-category-theory/Construction/Comma/Adjunction.v.html             │
│ 12.1600   2.8840   -9.2760  -76.28%   225  coq-category-theory/Construction/Comma/Adjunction.v.html             │
│ 10.7870   1.7730   -9.0140  -83.56%   289  coq-category-theory/Construction/Comma/Adjunction.v.html             │
│ 12.4980   3.8400   -8.6580  -69.28%   743  coq-category-theory/Construction/Comma/Adjunction.v.html             │
│ 23.3800  14.9280   -8.4520  -36.15%    93  coq-category-theory/Functor/Construction/Product/Monoidal.v.html     │
│ 13.4300   5.4370   -7.9930  -59.52%    49  coq-category-theory/Structure/Monoidal/Relevant.v.html               │
│ 10.2600   2.3550   -7.9050  -77.05%   226  coq-category-theory/Construction/Comma/Adjunction.v.html             │
│ 10.2140   2.4900   -7.7240  -75.62%   545  coq-category-theory/Construction/Comma/Adjunction.v.html             │
│  9.4760   1.8420   -7.6340  -80.56%   297  coq-category-theory/Construction/Comma/Adjunction.v.html             │
│ 10.2430   2.6730   -7.5700  -73.90%   579  coq-category-theory/Construction/Comma/Adjunction.v.html             │
│  9.0010   1.5700   -7.4310  -82.56%   214  coq-category-theory/Construction/Comma/Adjunction.v.html             │
│  8.7880   1.5330   -7.2550  -82.56%    70  coq-category-theory/Construction/Comma/Natural/Transformation.v.html │
│  8.1760   1.5220   -6.6540  -81.38%   215  coq-category-theory/Construction/Comma/Adjunction.v.html             │
│  9.0930   2.9830   -6.1100  -67.19%   676  coq-category-theory/Construction/Comma/Adjunction.v.html             │
│  8.4570   2.4800   -5.9770  -70.68%  1098  coq-category-theory/Functor/Construction/Product/Monoidal.v.html     │
│  8.4990   2.7660   -5.7330  -67.45%   871  coq-category-theory/Functor/Construction/Product/Monoidal.v.html     │
└─────────────────────────────────────────────────────────────────────────────────────────────────────────────────┘
```

</details>

**Comment by SkySkimmer** [comment]
coq-performance-tests-lite and coq-rewriter-perf-SuperFast failed with stack overflow

**Comment by SkySkimmer** [comment]
perf tests with -bt might show where the stack overflow comes from
https://gitlab.com/coq/coq/-/jobs/2246022795

**Comment by SkySkimmer** [comment]
@coqbot ci minimize

**Comment by SkySkimmer** [comment]
@coqbot ci minimize analysis

**Comment by SkySkimmer** [comment]
@coqbot ci minimize analysis

**Comment by JasonGross** [comment]
@coqbot CI minimize all

**Comment by JasonGross** [comment]
The issue is that ci-analysis also fails on master.  Coqbot should report the reason it suggests not minimizing targets when we request that it minimize them anyway, at least when there are no suggested targets.

**Comment by proux01** [comment]
That's bad, maybe @affeldt-aist want to hear about it (normally master is tested on analysis CI, so that shouldn't happen).

**Comment by proux01** [comment]
Analysis should now be fixed.

**Comment by affeldt-aist** [comment]
Thank you for letting me know.

**Comment by SkySkimmer** [comment]
The stack overflows in the bench jobs

**Comment by ppedrot** [comment]
OK, let's postpone.

**Comment by SkySkimmer** [comment]
This stack overflows on this PR but not on master with ulimit 4096 (on my machine)
~~~coq
Require Export Coq.Lists.List.
Require Export Coq.ZArith.ZArith.

Export ListNotations.

Definition Let_In {A P} (x : A) (f : forall a : A, P a) : P x := let y := x in f y.

Module Type AxT.
  Axiom Let_InAx : forall {A P} (x : A) (f : forall a : A, P a), P x.
  Axiom ZAddAx : Z -> Z -> Z.
End AxT.
Module Import Ax : AxT.
  Definition Let_InAx : forall {A P} (x : A) (f : forall a : A, P a), P x
    := @Let_In.
  Definition ZAddAx : Z -> Z -> Z
    := Z.add.
End Ax.


Eval lazy in
fun v : Z =>
nat_rect (fun _ : nat => (list Z -> list Z) -> list Z)
  (fun k : list Z -> list Z => k (repeat v (id Z.to_nat 9800%Z)))
  (fun (_ : nat) (rec : (list Z -> list Z) -> list Z) (k : list Z -> list Z) =>
   rec
     (fun ls : list Z =>
      list_rect (fun _ : list Z => (list Z -> list Z) -> list Z) (fun k0 : list Z -> list Z => k0 [])
        (fun (x : Z) (_ : list Z) (rec0 : (list Z -> list Z) -> list Z) (k0 : list Z -> list Z) =>
         Let_InAx (ZAddAx x x) (fun y : Z => rec0 (fun ys : list Z => k0 (y :: ys)))) ls k))
  (id Z.to_nat 1%Z) (fun x : list Z => x) = []
.
~~~

**Comment by SkySkimmer** [comment]
It produces a term like
~~~coq
fun v =>
       Let_InAx (ZAddAx v v)
         (fun y : Z =>
          Let_InAx (ZAddAx v v)
            (fun y0 : Z =>
             Let_InAx (ZAddAx v v)
               (fun y1 : Z =>
(* until y9000 *)
[ y; y0; ... ]
(* many closing parentheses *)
~~~
I guess we don't expect that to be produced in a tailrec way.
cc @JasonGross what do you expect from these tests? (rewrite_lift_lets_map)

**Comment by SkySkimmer** [comment]
ping @JasonGross

**Comment by Janno** [comment]
While we wait for @JasonGross, can I ask why this stack overflow comes up now? Is it just a matter of degree and the stack simply grows a bit faster with this MR? Or does the MR change the stack topology in a fundamental way? The diff looks very straight-forward so I would be surprised if it was the latter.

**Comment by SkySkimmer** [comment]
We don't know either.

**Comment by Janno** [comment]
Potentially useful data points: on my machine with 8.15.1, the test case overflows the a stack limited to 4396 kB but not one limited to 4496 kB (+100 kB).

This sounds to me like this test is already using a lot of stack space on master and the change simply pushed it over the edge.

(I find it interesting that the stack overflow occurs immediately (if it occurs at all). I suspect the code spawns all the recursive calls early on and then has to perform actual work only afterwards?)

**Comment by JasonGross** [comment]
Sorry for the delay, I somehow missed the tags.  This test is not meant to be testing stack but is instead testing performance of reduction vs various sorts of rewriting.  If we can fix the test by increasing the stack limit a little bit we should do that.

**Comment by SkySkimmer** [comment]
@coqbot bench

**Comment by SkySkimmer** [comment]
Looks like that works

**Review by ppedrot — APPROVED** [review]

**Comment by Alizter** [comment]
Nice work!

**Pull request merged**